### PR TITLE
qcontainer: Support pflash firmware files in qcow2 format

### PR DESCRIPTION
**Pull request by PaulYuuu**
By default, EDK2 provides firmware files in raw format, but fedora recently supported qcow2 format, and QEMU can native use it.

Reference: https://src.fedoraproject.org/rpms/edk2/c/3cf8c78157b21905db34f926f94a15ced66f52de?branch=rawhide
ID: 2196404

### CODE REVIEW

**Comment by PaulYuuu** [comment]
Hello @YongxueHong, can you help to review?

**Comment by PaulYuuu** [comment]
Hello @XueqiangWei, can you help run a regression test on x86? the only change is the file name from `*.fd` to `*.raw`.

**Comment by YongxueHong** [comment]
> Hello @XueqiangWei, can you help run a regression test on x86? the only change is the file name from `*.fd` to `*.raw`.

Hi @PaulYuuu 
What does the `fd` represent here?  is `file pflash`?

**Comment by PaulYuuu** [comment]
> > Hello @XueqiangWei, can you help run a regression test on x86? the only change is the file name from `*.fd` to `*.raw`.
> 
> Hi @PaulYuuu What does the `fd` represent here? is `file pflash`?

The file name, because we support 2 formats, so need to distinguish them. For x86, currently only support raw format, aarch64 supports both raw and qcow2.

**Comment by XueqiangWei** [comment]
> Hello @XueqiangWei, can you help run a regression test on x86? the only change is the file name from `*.fd` to `*.raw`.

Sure, run edk2 test loop with rhel9.3 on x86 platform, the results were passed.

The vars file name is changed to avocado-vt-vm1_rhel930-64-virtio-scsi-ovmf_qcow2_filesystem_VARS.raw in command lines.
e.g. 
-blockdev '{"node-name": "file_ovmf_vars", "driver": "file", "filename": "/root/avocado/data/avocado-vt/avocado-vt-vm1_rhel930-64-virtio-scsi-ovmf_qcow2_filesystem_VARS.raw"

By the way, could you please add "ID: ***" in Comment 0? Thanks.

**Comment by PaulYuuu** [comment]
> > Hello @XueqiangWei, can you help run a regression test on x86? the only change is the file name from `*.fd` to `*.raw`.
> 
> Sure, run edk2 test loop with rhel9.3 on x86 platform, the results were passed.
> 
> The vars file name is changed to avocado-vt-vm1_rhel930-64-virtio-scsi-ovmf_qcow2_filesystem_VARS.raw in command lines. e.g. -blockdev '{"node-name": "file_ovmf_vars", "driver": "file", "filename": "/root/avocado/data/avocado-vt/avocado-vt-vm1_rhel930-64-virtio-scsi-ovmf_qcow2_filesystem_VARS.raw"
> 
> By the way, could you please add "ID: ***" in Comment 0? Thanks.

OK, thanks for the reminder.

**Comment by PaulYuuu** [comment]
On aarch64:
```
# qcow2
[stdlog] 2023-05-15 22:31:44,493 avocado.test DEBUG|     aavmf_code_filename = QEMU_EFI-silent-pflash.qcow2
[stdlog] 2023-05-15 22:31:44,494 avocado.test DEBUG|     aavmf_vars_filename = vars-template-pflash.qcow2
[stdlog]     -blockdev '{"node-name": "file_aavmf_code", "driver": "file", "filename": "/usr/share/edk2/aarch64/QEMU_EFI-silent-pflash.qcow2", "auto-read-only": true, "discard": "unmap"}' \
[stdlog]     -blockdev '{"node-name": "drive_aavmf_code", "driver": "qcow2", "read-only": true, "file": "file_aavmf_code"}' \
[stdlog]     -blockdev '{"node-name": "file_aavmf_vars", "driver": "file", "filename": "/root/avocado/data/avocado-vt/avocado-vt-vm1_rhel930-aarch64-64k-virtio-scsi_qcow2_filesystem_VARS.qcow2", "auto-read-only": true, "discard": "unmap"}' \
[stdlog]     -blockdev '{"node-name": "drive_aavmf_vars", "driver": "qcow2", "read-only": false, "file": "file_aavmf_vars"}' \
```

```
# raw
[stdlog] 2023-05-15 22:31:44,493 avocado.test DEBUG|     aavmf_code_filename = QEMU_EFI-silent-pflash.raw
[stdlog] 2023-05-15 22:31:44,494 avocado.test DEBUG|     aavmf_vars_filename = vars-template-pflash.raw
  [stdlog]     -blockdev '{"node-name": "file_aavmf_code", "driver": "file", "filename": "/usr/share/edk2/aarch64/QEMU_EFI-silent-pflash.raw", "auto-read-only": true, "discard": "unmap"}' \
  [stdlog]     -blockdev '{"node-name": "drive_aavmf_code", "driver": "raw", "read-only": true, "file": "file_aavmf_code"}' \
  [stdlog]     -blockdev '{"node-name": "file_aavmf_vars", "driver": "file", "filename": "/root/avocado/data/avocado-vt/avocado-vt-vm1_rhel930-aarch64-64k-virtio-scsi_qcow2_filesystem_VARS.raw", "auto-read-only": true, "discard": "unmap"}' \
  [stdlog]     -blockdev '{"node-name": "drive_aavmf_vars", "driver": "raw", "read-only": false, "file": "file_aavmf_vars"}' \
```

**Review by YongxueHong — APPROVED** [review]
LGTM.

**Comment by YongxueHong** [comment]
Hi @XueqiangWei @zhenyzha 
Would you like to help to review it?
Thanks.

**Comment by XueqiangWei** [comment]
run edk2 test loop with rhel9.3 and win11 on x86 platform, the results were passed.
run qemu gating test, the results were passed.

The vars file name is changed to avocado-vt-vm1_rhel930-64-virtio-scsi-ovmf_qcow2_filesystem_VARS.raw in command lines.
e.g.
-blockdev '{"node-name": "file_ovmf_vars", "driver": "file", "filename": "/root/avocado/data/avocado-vt/avocado-vt-vm1_rhel930-64-virtio-scsi-ovmf_qcow2_filesystem_VARS.raw"

**Comment by XueqiangWei** [comment]
LGTM,
Acked-by: xuwei@redhat.com

**Review by XueqiangWei — APPROVED** [review]

File: virttest/qemu_devices/qcontainer.py [hunk 1]
```diff
@@ -1139,15 +1154,24 @@ def pflash_handler(firmware_name, cmd):
...
                         params.get("restore_%s_vars" % firmware_name) == "yes"):
                     shutil.copy2(pflash_vars_src_path, pflash_vars_path)
+
+                # Firmware vars file
                 if Flags.BLOCKDEV in self.caps:
                     protocol_pflash1 = qdevices.QBlockdevProtocolFile(pflash1)
-                    format_pflash1 = qdevices.QBlockdevFormatRaw(pflash1)
+                    if pflash_code_format == "raw":
+                        format_pflash1 = qdevices.QBlockdevFormatRaw(pflash1)
+                    elif pflash_code_format == "qcow2":
```
Comment (PaulYuuu) [inline virttest/qemu_devices/qcontainer.py:1169]: Here is a typo, should be `pflash_vars_format`, I will fix it asap.

**Pull request merged**